Desktop edge scrolling

justinmc opened the pull request:
When selection moves on desktop, the field will now scroll to view the scroll extent, like it already does on mobile.

![edge_mac_2](https://user-images.githubusercontent.com/389558/140590082-5dd94195-1e0d-4875-9799-523910e289c9.gif)

TODO:
 - [x] Make sure horizontal edge scrolling works.
 - [x] Make sure this works on Windows.
 - [x] Tests.

Fixes https://github.com/flutter/flutter/issues/32344
Fixes https://github.com/flutter/flutter/issues/59141

justinmc committed:
Works for Mac using same mechanism as iOS

justinmc committed:
Also reveal extent for keyboard selection changes

justinmc committed:
Do the same for other desktop platforms

justinmc committed:
Do the same for CupertinoTextField

justinmc committed:
Test and get platforms right in logic.

justinmc committed:
Vertical tests

justinmc commented:
@mdebbar Is this old TODO of yours actually fixed?

https://github.com/flutter/flutter/blob/2a00eacfacfed1f04d52346bd78ed02e37e39a19/packages/flutter/lib/src/widgets/text_selection.dart#L1530-L1531

I ran into a strange problem where vertical mouse dragging wasn't being detected properly in a test.  I'm wondering if there's anything unimplemented there that could be causing it, or if that TODO should be removed.

justinmc committed:
Update Cupertino logic to match Material

justinmc committed:
Cupertino test platforms correction

mdebbar commented:
@justinmc if you look at that code, it's only using a `HorizontalDragGestureRecognizer` which doesn't detect vertical dragging. At that time, vertical dragging was used for scrolling (even on web, because there was no scroll bars yet). But now I think that's not the case anymore. Maybe we should use `PanGestureRecognizer`?

justinmc committed:
Imprecise tests

justinmc committed:
Merge branch 'master' into edge-scrolling-desktop

justinmc committed:
More imprecise tests

justinmc commented:
Sounds good, thanks for the info.  I'll make a quick attempt at using PanGestureRecognizer and see if that fixes the strange missing vertical drag component that I saw in the test.  If it seems complicated, I'll save it for another PR.

justinmc committed:
Scrolling due to keyboard is already handled elsewhere

justinmc committed:
Try switching to PanGestureRecognizer

justinmc committed:
Keyboard tests don't work on web

justinmc committed:
Mouse vertical test now that PanGestureRecognizer allows it

justinmc committed:
Cupertino also shouldn't handle keyboard selection changes, handled elsewhere

justinmc commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1409,14 +1409,24 @@ class _TextSelectionGestureDetectorState extends State<TextSelectionGestureDetec
   DragStartDetails? _lastDragStartDetails;
   DragUpdateDetails? _lastDragUpdateDetails;
   Timer? _dragUpdateThrottleTimer;
+  bool _isMouseDragging = false;
```
HorizontalDragGestureRecognizer allows you to specify the type of pointer to accept, but now that we've switched to PanGestureRecognizer, which doesn't support that, we have to manually make sure we only listen to mouse drags.

LongCatIsLooong replied:
Would it be possible to add that to the existing `PanGestureRecognizer` or have a private subclass of `PanGestureRecognizer`?

justinmc replied:
Great idea, looks like it's actually only a 2 line change to add it to PanGesureRecognizer because the super class already supports it.

justinmc committed:
More test precision

justinmc commented:
@mdebbar Switching to PanGestureRecognizer worked without too much tweaking, and it fixed my test!  I've replaced it here in this PR and removed your TODO.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_field.dart:
```diff
@@ -967,15 +967,30 @@ class _CupertinoTextFieldState extends State<CupertinoTextField> with Restoratio
+    switch (defaultTargetPlatform) {
+      case TargetPlatform.iOS:
+      case TargetPlatform.macOS:
+        if (cause == SelectionChangedCause.longPress
```
Just curious what does a long-press do on iOS and macOS? Is it the force press floating cursor that makes them different?

justinmc replied:
Yup this handles the thing where you force press and then drag the cursor.

justinmc committed:
Add 'kind' support for PanGestureRecognizer instead of hacking it into text selection

LongCatIsLooong commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -614,7 +614,13 @@ class HorizontalDragGestureRecognizer extends DragGestureRecognizer {
+    PointerDeviceKind? kind,
+  }) : super(
+         debugOwner: debugOwner,
+         kind: kind,
```
`kind` is deprecated I think?

justinmc replied:
Ah I missed that, thanks.  I'll do the same thing but with `supportedDevices` as the deprecation message suggests.

justinmc committed:
Double check and standardize the test precision stuff.

justinmc committed:
Use the new supportedDevices instead of deprecated

LongCatIsLooong reviewed (approved):
LGTM. This should also close https://github.com/flutter/flutter/issues/28676 (the TODO issue) right?

justinmc commented:
@LongCatIsLooong That one is already closed, but I guess the TODO was missed.

justinmc commented:
Adding the cherry pick tag after discussing with @mariamhas, if we're still doing cherry picks for 2.8.  Shouldn't be a high risk of regressions.

devoncarew commented:
@justinmc and @HansMuller - when was this regression introduced? What's the impact if we don't take it for this hotfix? We should carefully weigh taking it now (the week right before the release) w/ delaying until a batch of fixes after the release.

justinmc commented:
I'm ok with skipping this.  The behavior without it is fairly bad, but it's actually not a regression.  Also there were previously a few goldens issues with this.  Probably best to let it be released normally if @mariamhas is alright with that.
